pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/PercentileKLLAggregationFunction.java

```diff
@@ -0,0 +1,248 @@
+import org.apache.pinot.core.query.aggregation.groupby.GroupByResultHolder;
+import org.apache.pinot.core.query.aggregation.groupby.ObjectGroupByResultHolder;
+import org.apache.pinot.segment.spi.AggregationFunctionType;
+import org.apache.pinot.spi.data.FieldSpec.DataType;
+
+/**
+ * <p>
+ *  {@code PercentileKLLAggregationFunction} provides an approximate percentile calculator using the KLL algorithm
+ *  from <a href="https://datasketches.apache.org/docs/KLL/KLLSketch.html">Apache DataSketches library</a>.
+ * </p>
+ * <p>
+ *  The interface is similar to plain 'Percentile' function except for the optional K value which determines
+ *  the size, hence the accuracy of the sketch.
+ * </p>
+ * <p><b>PERCENTILE_KLL(col, percentile, kValue)</b></p>
+ * <p>E.g.:</p>
+ * <ul>
+ *   <li><b>PERCENTILE_KLL(col, 90)</b></li>
+ *   <li><b>PERCENTILE_KLL(col, 99.9, 800)</b></li>
+ * </ul>
+ *
+ * <p>
+ *   There is a variation of the function (<b>PERCENTILE_RAW_KLL</b>) that returns the Base64 encoded
+ *   sketch object to be used externally.
+ * </p>
+ */
+public class PercentileKLLAggregationFunction
+    extends BaseSingleInputAggregationFunction<KllDoublesSketch, Comparable> {
+
+  protected final double _percentile;
+  protected int _kValue = 200; // size of the sketch. This is the default size used by DataSketches lib as well
+
+  public PercentileKLLAggregationFunction(List<ExpressionContext> arguments) {
+    super(arguments.get(0));
+
+    // Check that there are correct number of arguments
+    int numArguments = arguments.size();
+    Preconditions.checkArgument(numArguments == 2 || numArguments == 3,
+        "Expecting 2 or 3 arguments for PercentileKLL function: "
+            + "PERCENTILE_KLL(column, percentile, k=200");
+
+    _percentile = arguments.get(1).getLiteral().getDoubleValue();
+    if (numArguments == 3) {
+      _kValue = arguments.get(2).getLiteral().getIntValue();
+    }
+  }
+
+  public PercentileKLLAggregationFunction(ExpressionContext expression, double percentile) {
+    super(expression);
+    _percentile = percentile;
+  }
+
+  @Override
+  public AggregationFunctionType getType() {
+    return AggregationFunctionType.PERCENTILEKLL;
+  }
+
+  @Override
+  public AggregationResultHolder createAggregationResultHolder() {
+    return new ObjectAggregationResultHolder();
+  }
+
+  @Override
+  public GroupByResultHolder createGroupByResultHolder(int initialCapacity, int maxCapacity) {
+    return new ObjectGroupByResultHolder(initialCapacity, maxCapacity);
+  }
+
+  @Override
+  public void aggregate(int length, AggregationResultHolder aggregationResultHolder,
+      Map<ExpressionContext, BlockValSet> blockValSetMap) {
+    BlockValSet valueSet = blockValSetMap.get(_expression);
+    DataType valueType = valueSet.getValueType();
+    KllDoublesSketch sketch = getOrCreateSketch(aggregationResultHolder);
+
+    if (valueType == DataType.BYTES) {
```

Comment:
javadoc to explain the difference caused by valueType?

Reply:
Good suggestion. Updated the comments below and added explanation to function javadoc.